include/boost/math/special_functions/fibonacci.hpp

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020, Madhur Chauhan
+
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FIBO_HPP
+#define BOOST_MATH_SPECIAL_FIBO_HPP
+
+#include "boost/math/constants/constants.hpp"
+#include <boost/assert.hpp>
+#include <boost/math/policies/error_handling.hpp>
+#include <cmath>
+#include <limits>
+
+#ifdef _MSC_VER
+#pragma once
+#endif
+
+namespace boost {
+namespace math {
+
+template <typename T, class Policy>
+T fibonacci(unsigned long long n, const Policy &pol) {
+    using ull = unsigned long long;
+    BOOST_ASSERT(n >= 0);
+    if (n <= 3) return (n + 1) >> 1;
+    if (std::ceil(n * std::log2(boost::math::constants::phi<long double>()) - std::log2(5) / 2.0) > std::numeric_limits<T>::digits)
```

Comment:
I might do a bit of analysis to see if you could get away with `float` rather than `long double` here. I worry the check will take longer than the computation!

Reply:
I took the constants out from the `if` check. In future, I'm hoping if `std::log2` becomes `constexpr` then it would be very efficient. To deal with the imprecision issues with floating-point, I guess a difference of `std::numeric_limits<T>::epsilon` must be relaxed.

There are two extremes now, a very precise check or no check (in unchecked version).